Remove all orphan instances

treeowl opened the pull request:
* Remove all orphan instances.

* One of the `unions` definitions used `foldl`; make it use `foldl'`
  instead.

Closes #52.

treeowl commented:
@konsumlamm This PR moves a *lot* of things around, so I'm pretty much blocked on everything else till it goes in.

konsumlamm commented on src/Data/PQueue/Prio/Max.hs:
```diff
@@ -116,394 +112,5 @@ module Data.PQueue.Prio.Max (
--- can create thunk chains.
-seqSpine :: MaxPQueue k a -> b -> b
-seqSpine (MaxPQ q) = Q.seqSpine q
+import Prelude ()
```
What is this good for?

treeowl replied:
I doubt it's good for much. It was here before #26, which made it *mostly* redundant. `mapKeysMonotonic` suspends O(n) work. Under certain circumstances, you might want to force that work to happen immediately or at a particular time.

treeowl replied:
Oh, misread your comment. We don't want the `Prelude`, which exports various names that clash with ours.

konsumlamm replied:
Oh, right, didn't think there.

konsumlamm commented on src/Data/PQueue/Prio/Max.hs:
```diff
@@ -116,394 +112,5 @@ module Data.PQueue.Prio.Max (
-  traverse f (MaxPQ q) = MaxPQ <$> traverse f q
-  mapM = mapMWithKey . const
-  sequence = mapM id
-
--- | /O(1)/. Returns the empty priority queue.
-empty :: MaxPQueue k a
-empty = MaxPQ Q.empty
-
--- | /O(1)/. Constructs a singleton priority queue.
-singleton :: k -> a -> MaxPQueue k a
-singleton k a = MaxPQ (Q.singleton (Down k) a)
-
--- | Amortized /O(1)/, worst-case /O(log n)/. Inserts
--- an element with the specified key into the queue.
-insert :: Ord k => k -> a -> MaxPQueue k a -> MaxPQueue k a
-insert k a (MaxPQ q) = MaxPQ (Q.insert (Down k) a q)
```
Why did you move all the functions to `Data.PQueue.Prio.Max` instead of only the instances (and fucntions needed for the instances)?

treeowl replied:
It was much, much easier.

konsumlamm reviewed (commented):
Are the `Functor`, `Fodlable` and `Traversable` instances of `Down` used anywhere? Also, I think a better module name instead of `Data.PQueue.Prio.Max.Down` would be `Data.PQueue.Down`, since it's not specific to `MaxQueue` or `MaxPQueue`.

treeowl commented:
> Are the `Functor`, `Fodlable` and `Traversable` instances of `Down` used anywhere?

I don't know.

> Also, I think a better module name instead of `Data.PQueue.Prio.Max.Down` would be `Data.PQueue.Down`, since it's not specific to `MaxQueue` or `MaxPQueue`.

Good point.

treeowl commented on src/Data/PQueue/Prio/Max/Internals.hs:
```diff
@@ -1,23 +1,146 @@
 {-# LANGUAGE CPP #-}
 
-module Data.PQueue.Prio.Max.Internals where
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  Data.PQueue.Prio.Max
+-- Copyright   :  (c) Louis Wasserman 2010
+-- License     :  BSD-style
+-- Maintainer  :  libraries@haskell.org
+-- Stability   :  experimental
+-- Portability :  portable
+--
+-- General purpose priority queue.
+-- Each element is associated with a /key/, and the priority queue supports
+-- viewing and extracting the element with the maximum key.
+--
+-- A worst-case bound is given for each operation. In some cases, an amortized
+-- bound is also specified; these bounds do not hold in a persistent context.
+--
+-- This implementation is based on a binomial heap augmented with a global root.
+--
+-- We do not guarantee stable behavior.
+-- Ties are broken arbitrarily -- that is, if @k1 <= k2@ and @k2 <= k1@, then there
+-- are no guarantees about the relative order in which @k1@, @k2@, and their associated
+-- elements are returned. (Unlike Data.Map, we allow multiple elements with the
+-- same key.)
+--
+-- This implementation offers a number of methods of the form @xxxU@, where @U@ stands for
+-- unordered. No guarantees whatsoever are made on the execution or traversal order of
+-- these functions.
+-----------------------------------------------------------------------------
```
Uhhh...... Maybe I didn't need to do that.

treeowl replied:
Removed.

treeowl commented:
`Traversable Down` wasn't used, so I removed it. `Functor Down` and `Foldable Down` are currently used. We probably don't actually want to use `Foldable Down` (see #55), but we'll deal with that in a different PR. The `Functor` instance is used by things like `mapKeys` and `mapKeysMonotonic`. These are pretty inefficient no matter what we do (short of making all the trees lazy, which I'd very much rather not do), and I'm not optimistic that we can improve them much by reducing dictionary costs, but I guess we could try. I renamed the `Down` module `Data.PQueue.Internal.Down`, since it really is an internal thing.
